system/lib/wasmfs/syscalls.cpp

```diff
@@ -426,4 +430,79 @@ __wasi_errno_t __wasi_fd_seek(__wasi_fd_t fd,
 
   return __WASI_ERRNO_SUCCESS;
 }
+
+long __syscall_chdir(long path) {
+  auto pathParts = splitPath((char*)path);
+
+  if (pathParts.empty()) {
+    return -ENOENT;
+  }
+
+  long err;
+  auto dir = getDir(pathParts.begin(), pathParts.end(), err);
+
+  if (!dir) {
+    return err;
+  }
+
+  wasmFS.locked().setCWD(dir);
+  return 0;
+}
+
+long __syscall_getcwd(long buf, long size) {
+  // Check if buf points to a bad address.
+  if (!buf && size > 0) {
+    return -EFAULT;
+  }
+
+  // Check if the size argument is zero and buf is not a null pointer.
+  if (buf && size == 0) {
+    return -EINVAL;
+  }
+
+  auto curr = wasmFS.locked().getCWD();
+
+  std::string buffer = "";
+
+  while (curr != wasmFS.getRootDirectory()) {
+    auto currMaybeLocked = curr->maybeLocked();
+    // TODO: better error return value.
+    // Since we are traversing up the directory tree, this checks if another
+    // thread is trying to lock a common ancestor directory.
+    // If this is the case, getcwd will not contend for access and return.
```

Comment:
Can we safely retry rather than returning an error here?

Reply:
So in our discussion we talked about how we guarantee only one lock being held in `getDir`. So it looks like we won't need this try lock mechanism.